### PR TITLE
[API] Connect `import_function` to hub

### PR DESCRIPTION
**Pull request by yonishelach**
Until now `import_function` was connected to `mlrun/functions`, this PR connect it to the marketplace API.
Also now the default marketplace source is configured to be `mlrun/marketplace` github pages to support function versioning.
[ML-3370](https://jira.iguazeng.com/browse/ML-3370)

## Changes:
1. Added params to `list_sources` endpoint in order to search for a function easily in `import_function` flow (reduces number of api calls to one).
2. Removed `hub_url` parameter from `mlconf` and `client_spec` (doesn't really break BC).
3. Bumped `latest_data_version` to `4` in order to replace the old default marketplace to the new one.
4. Updated `marketplace.default_source` in config to the new one in `mlrun/marketplace` github pages.
5. Support versioning of functions in hub and added source name logic in url.
6. Since importing function from hub now requires a server side, unit tests that were do so moved to integration (or deleted if was already there)

### CODE REVIEW

**Comment by yonishelach** [comment]
Notice: backward compatibility test failed due to removing `hub_url` config from `client_spec`

File: mlrun/api/api/endpoints/marketplace.py [hunk 1]
```diff
@@ -78,7 +81,29 @@ async def list_sources(
...
+            for item in catalog.catalog:
+                if item.metadata.name == item_name:
+                    filtered_sources.append(source)
+                    break
+        return filtered_sources
+    elif tag or version:
+        raise mlrun.errors.MLRunBadRequestError(
+            "Tag or version params must be provided with item-name param"
+        )
+    return marketplace_sources
```
Comment (theSaarco) [inline mlrun/api/api/endpoints/marketplace.py:106]: Don't do this filtering here in the endpoint - please move it to the crud's `list_marketplace_sources` function (pass the parameters there and perform the filtering). The api endpoint should not do domain-specific functionality.

File: mlrun/api/initial_data.py [hunk 2]
```diff
@@ -494,33 +497,75 @@ def _enrich_project_state(
...
+        hub_marketplace_source = (
+            db_session.query(MarketplaceSource)
+            .filter(
+                MarketplaceSource.index
+                == mlrun.api.schemas.marketplace.last_source_index
+            )
+            .one_or_none()
         )
     except mlrun.errors.MLRunNotFoundError:
         hub_marketplace_source = None
```
Comment (theSaarco) [inline mlrun/api/initial_data.py:559]: This `try/execpt` is not needed anymore, since you changed the logic above to use a query with `one_or_none()` which will return `None` if the record doesn't exist.

File: mlrun/api/initial_data.py [hunk 3]
```diff
@@ -494,33 +497,75 @@ def _enrich_project_state(
...
+        db_session.delete(default_record)
+        db_session.commit()
+    else:
+        logger.info("Default marketplace source not found")
+
+
+def _update_default_marketplace_source(
+    db: mlrun.api.db.sqldb.db.SQLDB,
+    db_session: sqlalchemy.orm.Session,
+    overwrite: bool = False,
```
Comment (theSaarco) [inline mlrun/api/initial_data.py:521]: It seems that the `overwrite` parameter is pretty redundant - you call this function in 2 places:
1. as part of the data-migration flow, where `overwrite=True`
2. as part of initial-data flow, where it's `False`, but you only call this if there isn't a record already.
Note also that calling this with `overwrite=False` where a record already exists will result in an error/exception from the DB due to duplicate primary key value.

So, I think you should remove this parameter, and *always* delete the record if it exists (i.e. always call `_delete_default_marketplace_source`)

File: mlrun/api/initial_data.py [hunk 4]
```diff
@@ -494,33 +497,75 @@ def _enrich_project_state(
...
-        else:
-            logger.info("Not adding default marketplace source, per configuration")
-    return
+        _update_default_marketplace_source(db, db_session)
+
+
+def _perform_version_3_data_migrations(
+    db: mlrun.api.db.sqldb.db.SQLDB, db_session: sqlalchemy.orm.Session
+):
+    _update_default_marketplace_source(db, db_session, overwrite=True)
```
Comment (theSaarco) [inline mlrun/api/initial_data.py:568]: See my comment about the `overwrite` parameter.

File: mlrun/config.py [hunk 5]
```diff
@@ -80,7 +80,6 @@
     "default_project": "default",  # default project name
     "default_archive": "",  # default remote archive URL (for build tar.gz)
     "mpijob_crd_version": "",  # mpijob crd version (e.g: "v1alpha1". must be in: mlrun.runtime.MPIJobCRDVersions)
-    "hub_url": "https://raw.githubusercontent.com/mlrun/functions/{tag}/{name}/function.yaml",
```
Comment (theSaarco) [inline mlrun/config.py:83]: Did you verify that nobody overrides this using an env variable in deployments? As we discussed - if deployment overrides this parameter and it doesn't exist in the config, then you'll get an exception on initialization, so need to make sure nobody does that.

File: mlrun/db/httpdb.py [hunk 6]
```diff
@@ -2814,12 +2813,30 @@ def store_marketplace_source(
...
+        item_name: Optional[str] = None,
+        tag: Optional[str] = None,
+        version: Optional[str] = None,
+    ):
         """
         List marketplace sources in the MLRun DB.
+
+        :param item_name:   Sources contain this item will be returned, If not provided all sources will be returned.
+        :param tag:         Must be provided with item name, will be filtered by this tag.
+        :param version:     Must be provided with item name, will be filtered by this version.
```
Comment (theSaarco) [inline mlrun/db/httpdb.py:2827]: I think this docstring is very confusing - it is read as "if you provide an item name you must also provide tag AND version", which IIUC is really not the case. This eventually goes to `get_source_catalog` which supports getting just one of those, or none of them as well. I believe what you meant was something like "Supported only if item name if provided"

File: mlrun/api/api/endpoints/marketplace.py [hunk 7]
```diff
@@ -67,6 +67,9 @@ async def create_source(
     response_model=List[IndexedMarketplaceSource],
 )
 async def list_sources(
+    item_name: Optional[str] = Query(None),
+    tag: Optional[str] = Query(None),
+    version: Optional[str] = Query(None),
```
Comment (theSaarco) [inline mlrun/api/api/endpoints/marketplace.py:72]: Since `tag` and `version` are only supported when `item_name` is provided, then please verify that they are `None` if it wasn't provided, or explode.

File: mlrun/api/api/endpoints/marketplace.py [hunk 8]
```diff
@@ -67,6 +67,9 @@ async def create_source(
     response_model=List[IndexedMarketplaceSource],
 )
 async def list_sources(
+    item_name: Optional[str] = Query(None),
```
Comment (theSaarco) [inline mlrun/api/api/endpoints/marketplace.py:70]:
```suggestion
    item_name: Optional[str] = Query(None, alias="item-name"),
```
We only use dashes in REST parameter names.

File: mlrun/db/httpdb.py [hunk 9]
```diff
@@ -2814,12 +2813,30 @@ def store_marketplace_source(
...
+
+        :returns: List of indexed marketplace sources.
         """
         path = "marketplace/sources"
-        response = self.api_call(method="GET", path=path).json()
+        params = {}
+        if item_name:
+            params["item_name"] = normalize_name(item_name)
```
Comment (theSaarco) [inline mlrun/db/httpdb.py:2834]:
```suggestion
            params["item-name"] = normalize_name(item_name)
```

File: mlrun/run.py [hunk 10]
```diff
@@ -453,8 +453,8 @@ def import_function(url="", secrets=None, db="", project=None, new_name=None):
 
     special URLs::
 
-        function hub: hub://{name}[:{tag}]
-        local mlrun db:       db://{project-name}/{name}[:{tag}]
+        function hub:       hub://[{source}/]{name}[:{tag}]
```
Comment (theSaarco) [inline mlrun/run.py:456]: Just to make sure - the function name cannot contain `/` and it's verified elsewhere, right? Otherwise this is not deterministic.

File: mlrun/utils/helpers.py [hunk 11]
```diff
@@ -599,19 +599,44 @@ def generate_artifact_uri(project, key, tag=None, iter=None):
...
-    return config.get_hub_url().format(name=name, tag=tag), True
+    function_suffix = f"{name}/{tag}/src/function.yaml"
+    if not source_name:
+        # Searching item in all sources
+        sources = db.list_marketplace_sources(
+            item_name=normalize_name(name, verbose=False), tag=tag
+        )
+        if not sources:
+            raise mlrun.errors.MLRunNotFoundError(
+                f"Item={name}, tag={tag} not found in all marketplace sources"
```
Comment (theSaarco) [inline mlrun/utils/helpers.py:632]:
```suggestion
                f"Item={name}, tag={tag} not found in any marketplace source"
```

File: mlrun/utils/helpers.py [hunk 12]
```diff
@@ -599,19 +599,44 @@ def generate_artifact_uri(project, key, tag=None, iter=None):
...
+            )
+        # getting default source or first:
+        source = sources[-1] if sources[-1].index == -1 else sources[0]
```
Comment (theSaarco) [inline mlrun/utils/helpers.py:635]: This is incorrect. You're getting back the sources as they should be provided to the user - the default source is always last (to allow overriding by the user), so you should always return `sources[0]`.

**Review by theSaarco — CHANGES_REQUESTED** [review]
Overall looks good, but there are changes needed. Note that I didn't go over the tests yet.

File: mlrun/config.py (hunk 5 above)
Comment (yonishelach) [reply mlrun/config.py:83]: Sure, I checked it and indeed no one is using it.

File: mlrun/run.py (hunk 10 above)
Comment (yonishelach) [reply mlrun/run.py:456]: added a try except block for that

File: mlrun/db/httpdb.py [hunk 13]
```diff
@@ -2814,12 +2813,30 @@ def store_marketplace_source(
...
+        :param version:     Item version to filter by, supported only if item name is provided and tag is not.
+
+        :returns: List of indexed marketplace sources.
         """
         path = "marketplace/sources"
-        response = self.api_call(method="GET", path=path).json()
+        params = {}
+        if item_name:
+            params["item-name"] = normalize_name(item_name)
+        if item_name:
```
Comment (theSaarco) [inline mlrun/db/httpdb.py:2835]:
```suggestion
        if tag:
```

File: mlrun/db/httpdb.py [hunk 14]
```diff
@@ -2814,12 +2813,30 @@ def store_marketplace_source(
...
+            params["tag"] = tag
+        if item_name:
```
Comment (theSaarco) [inline mlrun/db/httpdb.py:2837]:
```suggestion
        if version:
```

File: tests/api/api/test_utils.py [hunk 15]
```diff
@@ -1307,10 +1307,18 @@ def _mock_import_function_to_dict(*args, **kwargs):
...
 
+    def _mock_extend_hub_uri_if_needed(*args, **kwargs):
+        return "some-url", True
+
     monkeypatch.setattr(
         mlrun.run, "import_function_to_dict", _mock_import_function_to_dict
     )
 
+    monkeypatch.setattr(
+        "mlrun.utils.helpers.extend_hub_uri_if_needed.__code__",
```
Comment (theSaarco) [inline tests/api/api/test_utils.py:1318]: Why is the `__code__` needed? You can just replace the function, can't you?

File: tests/rundb/test_httpdb.py [hunk 16]
```diff
@@ -670,88 +670,6 @@ def test_feature_vectors(create_server):
     ), "Features didn't get updated properly"
 
 
-def test_project_file_db_roundtrip(create_server):
```
Comment (theSaarco) [inline tests/rundb/test_httpdb.py:673]: Why is this test removed?

File: tests/projects/test_project.py [hunk 17]
```diff
@@ -587,29 +464,6 @@ def test_get_artifact_uri():
     assert uri == "store://models/arti/y:prod"
 
 
-def test_export_to_zip():
```
Comment (theSaarco) [inline tests/projects/test_project.py:590]: Why is this test removed? Is it because you're using a hub function? Can't you mock it?

**Review by theSaarco — CHANGES_REQUESTED** [review]
Functionality looks good, with some minor comments (looks like a copy-paste accident). 
Regarding tests - looks like you removed some tests that shouldn't be removed, or can at least remain with some mocking.

File: tests/rundb/test_httpdb.py (hunk 16 above)
Comment (yonishelach) [reply tests/rundb/test_httpdb.py:673]: Moved into integration under the name `test_project_db`

File: tests/projects/test_project.py (hunk 17 above)
Comment (yonishelach) [reply tests/projects/test_project.py:590]: edited: added mock

File: tests/rundb/test_httpdb.py (hunk 16 above)
Comment (yonishelach) [reply tests/rundb/test_httpdb.py]: added mocking now

**Review by theSaarco — APPROVED** [review]
LGTM :+1:

**Review by Tankilevitch — CHANGES_REQUESTED** [review]
minor stuff

File: mlrun/db/httpdb.py [hunk 18]
```diff
@@ -2858,12 +2857,30 @@ def store_hub_source(
         response = self.api_call(method="PUT", path=path, json=source)
         return mlrun.common.schemas.IndexedHubSource(**response.json())
 
-    def list_hub_sources(self):
+    def list_hub_sources(
+        self,
+        item_name: Optional[str] = None,
+        tag: Optional[str] = None,
+        version: Optional[str] = None,
+    ):
```
Comment (alonmr) [inline mlrun/db/httpdb.py:2865]: Add typehint for return value

File: mlrun/utils/helpers.py [hunk 19]
```diff
@@ -670,19 +670,49 @@ def generate_artifact_uri(project, key, tag=None, iter=None):
...
+                "Invalid character '/' in function name or source name"
+            )
     # hub function directory name are with underscores instead of hyphens
     name = name.replace("-", "_")
-    return config.get_hub_url().format(name=name, tag=tag), True
+    function_suffix = f"{name}/{tag}/src/function.yaml"
+    if not source_name:
+        # Searching item in all sources
+        sources = db.list_hub_sources(
+            item_name=normalize_name(name, verbose=False), tag=tag
```
Comment (alonmr) [inline mlrun/utils/helpers.py:704]: why normalize here and replace in line 699? you can normalize it once as I see it

File: mlrun/utils/helpers.py [hunk 20]
```diff
@@ -670,19 +670,49 @@ def generate_artifact_uri(project, key, tag=None, iter=None):
...
+        )
+        if not sources:
+            raise mlrun.errors.MLRunNotFoundError(
+                f"Item={name}, tag={tag} not found in any hub source"
+            )
+        # precedence to user source
+        source = sources[0]
+    else:
+        # Specific source is given
+        source = db.get_hub_source(source_name)
```
Comment (alonmr) [inline mlrun/utils/helpers.py:714]:
```suggestion
        indexed_source = sources[0]
    else:
        # Specific source is given
        indexed_source = db.get_hub_source(source_name)
```

**Review by alonmr — COMMENTED** [review]
Initial

File: mlrun/utils/helpers.py (hunk 19 above)
Comment (yonishelach) [reply mlrun/utils/helpers.py:715]: The normalize here is redundant, removing.

File: mlrun/api/initial_data.py [hunk 21]
```diff
@@ -537,48 +537,72 @@ def _enrich_project_state(
             db.store_project(db_session, project.metadata.name, project)
 
 
+def _delete_default_hub_source(db_session: sqlalchemy.orm.Session):
```
Comment (alonmr) [inline mlrun/api/initial_data.py:540]: IMO it's easier to read code top to bottom, meaning that the called method is below the caller.
It's easier to understand the flow without having to guess what's the end goal of the helper methods.
It seems that this file is written the opposite so if it's on purpose we can discuss but if not, please flip the order.

File: mlrun/utils/helpers.py [hunk 22]
```diff
@@ -681,19 +681,48 @@ def generate_artifact_uri(project, key, tag=None, iter=None):
...
-        loc = name.find(":")
-        tag = name[loc + 1 :]
-        name = name[:loc]
+def extend_hub_uri_if_needed(uri) -> Tuple[str, bool]:
+    """
+    Retrieve the full uri of the item's yaml in the hub.
+
+    :param uri: structure: "hub://[<source>/]<item-name>[:<version>]"
 
+    :return: (extended uri of item, is hub item)
```
Comment (alonmr) [inline mlrun/utils/helpers.py:690]:
```suggestion
    :return: A tuple of:
               [0] = Extended URI of item
               [1] =  Is hub item (bool)
```

File: mlrun/utils/helpers.py [hunk 23]
```diff
@@ -681,19 +681,48 @@ def generate_artifact_uri(project, key, tag=None, iter=None):
...
-    tag = "master"
-    if ":" in name:
-        loc = name.find(":")
-        tag = name[loc + 1 :]
-        name = name[:loc]
+def extend_hub_uri_if_needed(uri) -> Tuple[str, bool]:
+    """
+    Retrieve the full uri of the item's yaml in the hub.
+
+    :param uri: structure: "hub://[<source>/]<item-name>[:<version>]"
```
Comment (alonmr) [inline mlrun/utils/helpers.py:688]:
```suggestion
    :param uri: structure: "hub://[<source>/]<item-name>[:<tag>]"
```

File: tests/api/api/test_submit.py [hunk 24]
```diff
@@ -183,8 +195,11 @@ def test_submit_job_ensure_function_has_auth_set(
 
 
 def test_submit_schedule_job_from_hub_from_ui(
-    db: Session, client: TestClient, pod_create_mock, k8s_secrets_mock
+    db: Session, client: TestClient, pod_create_mock, k8s_secrets_mock, monkeypatch
 ) -> None:
+    monkeypatch.setattr(
+        mlrun.run, "extend_hub_uri_if_needed", _mock_extend_hub_uri_if_needed
```
Comment (alonmr) [inline tests/api/api/test_submit.py:201]: better mock `list_hub_sources` or `get_hub_source` if needed but the helper is part of what the test is testing

File: tests/api/api/test_utils.py [hunk 25]
```diff
@@ -1307,10 +1307,17 @@ def _mock_import_function_to_dict(*args, **kwargs):
...
         return original_function
 
+    def _mock_extend_hub_uri_if_needed(*args, **kwargs):
+        return "some-url", True
+
     monkeypatch.setattr(
         mlrun.run, "import_function_to_dict", _mock_import_function_to_dict
     )
 
+    monkeypatch.setattr(
```
Comment (alonmr) [inline tests/api/api/test_utils.py:1317]: here it's ok to mock because it's not part of what the test is testing

File: tests/api/utils/test_builder.py [hunk 26]
```diff
@@ -420,6 +421,7 @@ def test_resolve_mlrun_install_command_version():
 
 def test_build_runtime_ecr_with_ec2_iam_policy(monkeypatch):
     _patch_k8s_helper(monkeypatch)
+    _mock_import_hub_function(monkeypatch)
```
Comment (alonmr) [inline tests/api/utils/test_builder.py:424]: this is not related to the test, just change the `set_function` to not import a hub function

File: tests/utils/test_helpers.py [hunk 27]
```diff
@@ -150,46 +149,6 @@ def test_spark_job_name_regex(value, expected):
...
-            "expected_output": "https://raw.githubusercontent.com/mlrun/functions/master/function_name/function.yaml",
-        },
-        {
-            "input_uri": "hub://function-name:development",
-            "expected_output": "https://raw.githubusercontent.com/mlrun/functions/development/function_name/function.ya"
-            "ml",
-        },
-    ],
-)
-def test_extend_hub_uri(case):
```
Comment (alonmr) [inline tests/utils/test_helpers.py:180]: try to fix the test if you can to keep coverage

File: tests/rundb/test_httpdb.py [hunk 28]
```diff
@@ -738,7 +738,6 @@ def test_project_file_db_roundtrip(create_server):
     function_name = "trainer-function"
     function = mlrun.new_function(function_name, project_name)
     project.set_function(function, function_name)
-    project.set_function("hub://describe", "describe")
```
Comment (alonmr) [inline tests/rundb/test_httpdb.py:741]: why remove? this is an integration test I think it should work

File: tests/common_fixtures.py [hunk 29]
```diff
@@ -485,6 +486,22 @@ def _get_function_internal(self, function_name: str = None):
     def store_metric(self, uid, project="", keyvals=None, timestamp=None, labels=None):
         pass
 
+    def list_hub_sources(self, item_name: str = None, tag: str = None):
```
Comment (alonmr) [inline tests/common_fixtures.py:489]: Is this used?

**Review by alonmr — COMMENTED** [review]
Great job!
A few more comment

File: mlrun/api/crud/hub.py [hunk 30]
```diff
@@ -310,3 +322,41 @@ def get_asset(
...
+        for source in sources:
+            catalog = self.get_source_catalog(
+                source=source.source,
+                version=version,
+                tag=tag,
+            )
+            for item in catalog.catalog:
+                if item.metadata.name == item_name:
+                    filtered_sources.append(source)
+                    break
```
[the remark this answers is not in the document]
Comment (yonishelach) [reply mlrun/api/crud/hub.py:361]: Notice this is a nested for loop, so when the item is found at least once in the catalog it means that the source is appended to the list of sources and there is no need to search further in this catalog - the break only breaks the inner loop.
Therefore, **ALL** the sources that contains this item in their catalog will be returned.

File: tests/common_fixtures.py (hunk 29 above)
Comment (yonishelach) [reply tests/common_fixtures.py:489]: No, removing

Comment (yonishelach) [reply tests/common_fixtures.py]: In the end I am using it - added `rundb_mock` in `test_helpers.py`

**Review by alonmr — APPROVED** [review]
